dosinst: Use $HOMEDRIVE$HOMEPATH and $USERPROFILE as fallbacks of plugin directories

k-takata opened the pull request:
Creating plugin directories in $VIM as a fallback of $HOME is not useful.

If the install.exe is executed from the NSIS installer, $VIM is normally
`C:\Program Files (x86)\Vim`.  Normal users cannot put plugin files into it,
because the directory requires administrator privileges.

Currently, the NSIS installer has the following two options:
* [x] Create plugin directories in HOME or VIM
* [ ] Create plugin directories in VIM

The first one is the default, however, most people might not set $HOME.
Then the plugin directories will be created in $VIM (which is not useful).
If the user really wants to create in $VIM, he should select the second one.

If the user doesn't set $HOME, it should fallback to $USERPROFILE.
What do you think?

k-takata committed:
dosinst: Use $USERPROFILE as a fallback of plugin directories

Creating plugin directories in $VIM as a fallback of $HOME is not useful.

If the install.exe is executed from the NSIS installer, $VIM is normally
`C:\Program Files (x86)\Vim`.  Normal users cannot put plugin files into
it, because the directory requires administrator privileges.

Currently, the NSIS installer has the following two options:
* [x] Create plugin directories in HOME or VIM
* [ ] Create plugin directories in VIM

The first one is the default, however, most people might not set $HOME.
Then the plugin directories will be created in $VIM (which is not useful).
If the user really wants to create in $VIM, he should select the second
one.

If the user doesn't set $HOME, it should fallback to $USERPROFILE.

h-east commented:
Yeah, This change makes sense 👍

brammool commented:
Ken Takata wrote:

> Creating plugin directories in $VIM as a fallback of $HOME is not useful.
> 
> If the install.exe is executed from the NSIS installer, $VIM is normally
> `C:\Program Files (x86)\Vim`.  Normal users cannot put plugin files into it,
> because the directory requires administrator privileges.
> 
> Currently, the NSIS installer has the following two options:
> * [x] Create plugin directories in HOME or VIM
> * [ ] Create plugin directories in VIM
> 
> The first one is the default, however, most people might not set $HOME.
> Then the plugin directories will be created in $VIM (which is not useful).
> If the user really wants to create in $VIM, he should select the second one.
> 
> If the user doesn't set $HOME, it should fallback to $USERPROFILE.
> What do you think?

We should use the same mechanism as in init_homedir().  We could even
move the shared logic to a separate file, so that they stay in sync.

If HOMEDRIVE / HOMEPATH differs from USERPROFILE, then the installer
would put files in USERPROFILE, while Vim would look in HOMEDRIVE /
HOMEPATH.  That's probably not working well.

If $VIM is starting with "C:\Program Files" we should probably not
suggest the user to put plugins there.  It's only useful on a multi-user
machine where you intentionally want to share plugins.  I think the user
needs to do that manually, it's too complicated for the installer.
And it may trick the user in deleting all his plugins when he upgrades
to a newer version.

Argh, so many things to keep in mind.

-- 
MAN:     You don't frighten us, English pig-dog!  Go and boil your bottoms,
         son of a silly person.  I blow my nose on you, so-called Arthur-king,
         you and your silly English K...kaniggets.
   He puts hands to his ears and blows a raspberry.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

k-takata committed:
dosinst: Use also $HOMEDRIVE$HOMEPATH for fallback

Copy init_homedir() from misc1.c and modified for dosinst.
It does:
1. Get $HOME.
2. If not found, get $HOMEDRIVE$HOMEPATH.
3. If not found, get $USERPROFILE.
4. Expand %var% if exists.

k-takata commented:
Okay, I copied (and modified) the code from init_homedir() in misc1.c. Sharing the code might not work well, because it uses `vim_*` functions (e.g. `vim_strncpy`) which are not available in dosinst.c.
So, now dosinst.c uses the same mechanism as in init_homedir() in misc1.c:

1. Get `$HOME`.
2. If not found, get `$HOMEDRIVE$HOMEPATH`.
3. If not found, get `$USERPROFILE`.
4. Expand `%var%` if exists.

k-takata committed:
dosinst: Fix that installfunc is initialized twice

text and arg were not initialized.

k-takata committed:
dosinst: Fix that `-create-directories` couldn't used twice

If a user selects following two components in the NSIS installer:

* Create plugin directories in HOME or VIM
* Create plugin directories in VIM

dosinst.c will be called with the following two command line options:
`-create-directories home -create-directories vim`.

dosinst.c couldn't handle them. `home` was overwritten by `vim` and
it created the directories only in $VIM.

Remember the both options and create both in $HOME and $VIM.

k-takata commented:
I found another problem in dosinst.c.

If a user selects following two components in the NSIS installer:

* Create plugin directories in HOME or VIM
* Create plugin directories in VIM

dosinst.c will be called with the following two command line options:
`-create-directories home -create-directories vim`.

However, dosinst.c couldn't handle them. `home` was overwritten by `vim` and
it created the directories only in $VIM.

I added a commit to fix this.

k-takata commented:
@brammool I think this is ready to be merged. Do you have any concern?

brammool commented:
Ken Takata wrote:

> @brammool I think this is ready to be merged. Do you have any concern?

I'll look into it soon.

-- 
MORTICIAN:    Bring out your dead!
              [clang]
              Bring out your dead!
              [clang]
              Bring out your dead!
CUSTOMER:     Here's one -- nine pence.
DEAD PERSON:  I'm not dead!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
